Support pipeline `secrets` syntax

matthewborden opened the pull request:
This adds a new secrets field to command steps. This allows using Buildkite Secrets in `pipeline.yml`. Secrets can be included at both the pipeline level and step level, with step secrets taking precedence.

Basic usage:
```yaml
steps:
  - command: "deploy.sh"
     secrets:
       - DATABASE_URL
       - API_TOKEN
```

Pipeline and step secrets merge together:
```yaml
secrets:
  - DATABASE_URL
  - REDIS_URL

steps:
  - command: "test.sh"
    secrets:
      - API_TOKEN        # Step-specific secret
      - DATABASE_URL

  - command: "deploy.sh" # Includes (DATABASE_URL, REDIS_URL, API_TOKEN)
```

For now this only supports the simple string array format where the secret key becomes the environment variable name. The Secret struct is designed to support future extensions like custom environment variable names, file secrets, SSH keys, etc.

Secrets are included in signed pipelines when then the step has secrets.

matthewborden committed:
Support pipeline `secrets` syntax

This adds a new secrets field to command steps that works similarly to how
plugins are handled. Secrets can be defined at both the pipeline level
and step level, with step secrets taking precedence.

Basic usage:
```yaml
secrets:
  - DATABASE_URL
  - API_TOKEN
steps:
  - command: "deploy.sh"
```

Pipeline and step secrets merge together:
```yaml
secrets:
  - DATABASE_URL    # Available to all steps
  - REDIS_URL

steps:
  - command: "test.sh"
    secrets:
      - API_TOKEN        # Step-specific secret
      - DATABASE_URL     # Overrides pipeline-level DATABASE_URL

  - command: "deploy.sh" # Inherits only pipeline secrets
```

For now this only supports the simple string array format where the secret
key becomes the environment variable name. The Secret struct is designed
to support future extensions like custom environment variable names, file
secrets, SSH keys, etc.

Secrets are included in signed pipelines when then the step has secrets.

DrJosh9000 commented on signature/pipeline_invariants_test.go:
```diff
@@ -0,0 +1,179 @@
+
+	if secrets[0].Key != "DATABASE_URL" {
+		t.Errorf("SignedFields() expected secret key 'DATABASE_URL', got %q", secrets[0].Key)
+	}
```
[`cmp.Diff`](https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff) can replace these four separate checks (and check `EnvironmentVariable` too):


```suggestion
	want := pipeline.Secrets{
		{Key: "DATABASE_URL", EnvironmentVariable: &dbUrl},
	}
	if diff := cmp.Diff(fields["secrets"], want); diff != "" {
		t.Errorf("step.SignedFields()[\"secrets\"] diff (-got +want):\n%s", diff)
	}
```

DrJosh9000 commented on signature/pipeline_invariants_test.go:
```diff
@@ -0,0 +1,179 @@
+			Command: "echo hello",
+			Env:     map[string]string{"FOO": "bar"},
+			Plugins: nil, // nil plugins - this should become nil
+			Secrets: nil, // nil secrets - this should become nil
```
Comments here seem strange - wouldn't we expect `nil` to _remain_ `nil`?

```suggestion
			Plugins: nil,
			Secrets: nil,
```

DrJosh9000 commented on signature/pipeline_invariants_test.go:
```diff
@@ -0,0 +1,179 @@
+
+	fields, err := step.SignedFields()
+	if err != nil {
+		t.Fatalf("SignedFields() error = %v", err)
```
[Optional thoughts] Perhaps we should make clear that `SignedFields` is a method of `step`, e.g.

```suggestion
		t.Fatalf("step.SignedFields() error = %v", err)
```

and maybe even that `step` has the type `commandStepWithInvariants`, but there's probably enough context to figure that out already.

DrJosh9000 commented on signature/pipeline_invariants_test.go:
```diff
@@ -0,0 +1,179 @@
+	// Check that secrets field is NOT present when empty (for backward compatibility)
+	if _, has := fields["secrets"]; has {
+		t.Errorf("SignedFields() should not include 'secrets' field when empty for backward compatibility")
+	}
```
For messages where there are no formatting tokens (`%v`, `%q`, etc) there are `f`-less methods, `t.Error` and `t.Fatal`:

```suggestion
	// Check that secrets field is NOT present when empty (for backward compatibility)
	if _, has := fields["secrets"]; has {
		t.Error("step.SignedFields() should not include 'secrets' field when empty for backward compatibility")
	}
```

DrJosh9000 commented on signature/pipeline_invariants_test.go:
```diff
@@ -0,0 +1,179 @@
+
+	if secrets[0].Key != "DATABASE_URL" {
+		t.Errorf("ValuesForFields() expected secret key 'DATABASE_URL', got %q", secrets[0].Key)
+	}
```
Similarly to the above, `cmp.Diff` could save a bit of code here:

```suggestion
	want := pipeline.Secrets{
		{Key: "DATABASE_URL", EnvironmentVariable: &dbUrl},
	}
	if diff := cmp.Diff(values["secrets"], want); diff != "" {
		t.Errorf("step.ValuesForFields(%v)[\"secrets\"] diff (-got +want):\n%s", fields, diff)
	}
```

DrJosh9000 commented on signature/pipeline_invariants_test.go:
```diff
@@ -0,0 +1,179 @@
+	fields := []string{"command", "env", "plugins", "matrix", "repository_url", "secrets"}
+	values, err := step.ValuesForFields(fields)
+	if err != nil {
+		t.Fatalf("ValuesForFields() error = %v", err)
```
The input to the function could be relevant to the failure, so it's important to include it wherever reasonable:

```suggestion
		t.Fatalf("step.ValuesForFields(%v) error = %v", fields, err)
```

DrJosh9000 commented on signature/pipeline_invariants_test.go:
```diff
@@ -0,0 +1,179 @@
+	expectedError := "one or more required fields are not present: [secrets]"
+	if err.Error() != expectedError {
+		t.Errorf("ValuesForFields() expected error %q, got %q", expectedError, err.Error())
+	}
```
Good use of `%q` 👍

Checking the error string is acceptable since we control the implementation, but is prone to being a "change detector". At some point maybe we should introduce a new error type that can be [checked more semantically](https://google.github.io/styleguide/go/decisions#test-error-semantics) than  we can via a string comparison. 

Other comment here is ["got before want"](https://google.github.io/styleguide/go/decisions#got-before-want):

```suggestion
	wantErr := "one or more required fields are not present: [secrets]"
	if err.Error() != wantErr {
		t.Errorf("ValuesForFields() error = %q, want %q", err.Error(), wantErr)
	}
```

DrJosh9000 commented on signature/pipeline_invariants_test.go:
```diff
@@ -0,0 +1,179 @@
+	// Should not have secrets field
+	if _, has := values["secrets"]; has {
+		t.Errorf("ValuesForFields() should not include 'secrets' field when step has no secrets")
+	}
```
The focus here is on the secrets field, but perhaps we should diff the whole return value anyway?

```suggestion
	wantValues := map[string]any{
		"command": "echo hello",
		"env": map[string]string{"FOO": "bar"},
		"plugins": nil,
		"matrix": nil,
		"repository_url", "https://github.com/example/repo",
	}
	if diff := cmp.Diff(values, wantValues); diff != "" {
		t.Errorf(step.ValuesForFields(%v) diff (-got +want):\n%s", fields, diff)
	}
```

DrJosh9000 reviewed (commented):
First pass of comments

DrJosh9000 commented on secret.go:
```diff
@@ -0,0 +1,53 @@
+func (s *Secret) MarshalJSON() ([]byte, error) {
+	type secretAlias Secret
+	return json.Marshal((*secretAlias)(s))
+}
```
This pretty much implements the default behaviour. 
The main reason to provide a custom `MarshalJSON` implementation here would be to support the `"inline"`-tagged field (`RemainingFields`), which we have a pre-written function for:


```suggestion
// MarshalJSON marshals the secret to JSON.
func (s *Secret) MarshalJSON() ([]byte, error) {
	return inlineFriendlyMarshalJSON(s)
}
```

(`"inline"` is a `yaml`-package convention, but not one supported by the stdlib `encoding/json`).

If we don't need to future-proof the struct (i.e. we declare right now that `key` and `environment_variable` are the only attributes, or that if we decide to support them, they require e.g. an agent upgrade), then we don't need `RemainingFields` or this custom marshaler method at all.

DrJosh9000 commented on secret.go:
```diff
@@ -0,0 +1,53 @@
+func (s *Secret) MarshalYAML() (any, error) {
+	type secretAlias Secret
+	return (*secretAlias)(s), nil
+}
```
This `MarshalYAML` also implements the default behaviour, so it could be omitted (along with the check for the `yaml.Marshaler` implementation on line 13).

DrJosh9000 commented on secret.go:
```diff
@@ -0,0 +1,53 @@
+// Secret represents a pipeline secret configuration.
+type Secret struct {
+	Key                 string         `json:"key" yaml:"key"`
+	EnvironmentVariable *string        `json:"environment_variable,omitempty" yaml:"environment_variable,omitempty"`
```
I wonder if this should be made `string` instead of `*string`? 

The empty string is treated as "empty" for the purposes of `omitempty`. The empty string is also definitely _not_ a valid environment variable name, so seems to me that there's no benefit to distinguishing between "not provided" and "empty but provided".

```suggestion
	EnvironmentVariable string        `json:"environment_variable,omitempty" yaml:"environment_variable,omitempty"`
```

matthewborden replied:
I'd like to be able to support a future where we extend secrets with other types (ssh keys, files etc), and by having the environment variable by nullable, we could return a type of `{ key: "SECRET_NAME", file: {...} }` from the API as `step`. If it's not nullable, we'd need to return `{ key: "SECRET_NAME", file: {...}, environment_variable: "" }` which while possible, doesn't seem accurate or within our normal API conventions. Does that make sense to continue using `*string`?

DrJosh9000 replied:
If an attribute is "missing" from the JSON or YAML, then the unmarshaler won't return an error, it just won't fill in the field. So, the API wouldn't need to include `environment_variable: ""`. `omitempty` takes care of the other direction (skipping the attribute in the output when it is empty).

matthewborden replied:
> If an attribute is "missing" from the JSON or YAML, then the unmarshaler won't return an error, it just won't fill in the field.

Will that mean that `environment_variable` will be `nil` or "" when not set by the API?

DrJosh9000 replied:
It would be left at the default value for the type - for `*string` that's `nil`, for `string` that is `""`.

matthewborden replied:
Thanks!

DrJosh9000 commented on secrets.go:
```diff
@@ -0,0 +1,153 @@
+				}
+				*s = append(*s, secret)
+
+			case map[string]any:
```
I don't think `UnmarshalOrdered` will ever be passed a slice of `map[string]any`. The `ordered` package exists to assume all YAML maps are ordered, and then we lower into these non-ordered map types when "unmarshaling" from ordered types.

DrJosh9000 commented on secrets.go:
```diff
@@ -0,0 +1,153 @@
+					if key, ok := keyVal.(string); ok {
+						secret.Key = key
+					}
+				}
```
This logic might benefit from inverting. Given that `key` has to be present and must be a string, we could return early. Something like:

```suggestion
				keyVal, _ := ct.Get("key")
				key, ok := keyVal.(string)
				if !ok {
					return fmt.Errorf("unmarshaling secret: key must be a string, but was %T", keyVal)
				}
				secret.Key = key
```

And `key` must be non-empty, right? Which means the "default zero value"  property of various expressions in Go can be used to check for missing, empty, and incorrect type:

 ```suggestion
 				keyVal, _ := ct.Get("key")
 				// Still need two values on the left side to avoid panic
				key, _ := keyVal.(string)
				if key == "" {
					return fmt.Errorf("unmarshaling secret: key must be a non-empty string, but was %[1]T %[1]q", keyVal)
				}
				secret.Key = key
```

which would subsume the check below for `secret.Key == ""`.

moskyb replied:
ohhh, TIL `%[1]<printf directive>` for only having to pass one value to printf 👀

DrJosh9000 commented on secrets.go:
```diff
@@ -0,0 +1,153 @@
+					if envVar, ok := envVarVal.(string); ok {
+						secret.EnvironmentVariable = &envVar
+					}
+				}
```
Similarly it might be better UX to report if the user wrote a non-string given YAML sometimes parses things non-obviously:


```suggestion
				if envVarVal, _ := ct.Get("environment_variable"); envVarVal != nil {
					envVar, ok := envVarVal.(string)
					if !ok {
						return fmt.Errorf("unmarshaling secret: environment_variable must be a string, but was %T", envVarVal)
					}
					secret.EnvironmentVariable = &envVar
				}		
```

DrJosh9000 reviewed (commented):
Second pass

matthewborden committed:
Go formatting and review comments

matthewborden committed:
Use `cmp.Diff` for test comparisons

matthewborden committed:
Switch to inlineFriendlyMarshalJSON for Secret marshaling

DrJosh9000 commented on secret_test.go:
```diff
@@ -0,0 +1,79 @@
+		EnvironmentVariable: &envVar,
+	}
+
+	want := `{"key":"DATABASE_URL","environment_variable":"DATABASE_URL","RemainingFields":null}`
```
Inline processing would mean `"RemainingFields":...` shouldn't be present in the output. Also I believe the JSON marshaler sorts the attributes.
 
```suggestion
	want := `{"environment_variable":"DATABASE_URL","key":"DATABASE_URL"}`
```

matthewborden replied:
```
    secret_test.go:27: json.Marshal(pipeline.Secret{Key:"DATABASE_URL", EnvironmentVariable:"DATABASE_URL", RemainingFields:map[string]interface {}(nil)}) = "{\"key\":\"DATABASE_URL\",\"environment_variable\":\"DATABASE_URL\",\"RemainingFields\":null}", want "{\"environment_variable\":\"DATABASE_URL\",\"key\":\"DATABASE_URL\"}"
```

The test I've written `TestSecretMarshalJSON` is still including `RemainingFields` when using `inlineFriendlyMarshalJSON`. Do I need to add further tags to prevent `RemainingFields` from being included?

DrJosh9000 replied:
Hmm, seems a bit weird. This looks the same as the other uses of `inlineFriendlyMarshalJSON`...

DrJosh9000 replied:
Ah, the `MarshalJSON` method is on the type `*Secret`, but you're passing a `Secret`. This must be preventing the stdlib JSON package from noticing the `MarshalJSON` method. 
A little bit of `log.Print`-debugging confirms.

The solution: either change `MarshalJSON` like so:

```diff
 // MarshalJSON marshals the secret to JSON.
-func (s *Secret) MarshalJSON() ([]byte, error) {
+func (s Secret) MarshalJSON() ([]byte, error) {
 	return inlineFriendlyMarshalJSON(s)
 }
```

or change the test to use a pointer to a `Secret`:

```diff
-	secret := Secret{
+ 	secret := &Secret{
```

Elsewhere I guess we must just be sticking to pointers everywhere.

matthewborden committed:
pipeline.Secret.EnvironmentVariable string

matthewborden committed:
Use cmp.Diff for test assertions

matthewborden committed:
Remove additional pointer Secret usage

matthewborden committed:
Update secrets.go

Co-authored-by: Josh Deprez <jd@buildkite.com>

matthewborden committed:
fmt

matthewborden committed:
Check environment variable type is correct

matthewborden committed:
MarshalJSON should be pointer-less

Defining the method with a pointer caused stdlib JSON not to detect it

// MarshalJSON marshals the secret to JSON.
-func (s *Secret) MarshalJSON() ([]byte, error) {
+func (s Secret) MarshalJSON() ([]byte, error) {
 	return inlineFriendlyMarshalJSON(s)
 }

matthewborden committed:
JSON unmarshaling requires a pointer to Secret

matthewborden committed:
More removing pointers for Secret.EnvironmentVariable

matthewborden committed:
Typed nils for matrix and plugins

matthewborden committed:
Add test for signing secrets

DrJosh9000 commented on pipeline_secrets_test.go:
```diff
@@ -0,0 +1,118 @@
+		t.Fatalf("ordered.Unmarshal() error = %v", err)
+	}
+
+	if len(p.Secrets) != 2 {
+		t.Fatalf("len(p.Secrets) = %d, want 2", len(p.Secrets))
+	}
```
This is adequately covered by the `cmp.Diff` check below.

DrJosh9000 commented on pipeline_secrets_test.go:
```diff
@@ -0,0 +1,118 @@
+	step1.MergeSecretsFromPipeline(p.Secrets)
+	step2.MergeSecretsFromPipeline(p.Secrets)
+
+	// Step1 should have 3 secrets (2 from step + 1 from pipeline not overridden)
+	if len(step1.Secrets) != 3 {
+		t.Fatalf("len(step1.Secrets after merge) = %d, want 3", len(step1.Secrets))
+	}
+
+	// Step2 should have 2 secrets (both from pipeline)
+	if len(step2.Secrets) != 2 {
+		t.Fatalf("len(step2.Secrets after merge) = %d, want 2", len(step2.Secrets))
+	}
```
This part at least would be a good candidate for full structure comparisons (`cmp.Diff`) - to be able to say that the not-overridden secret actually wasn't overridden, and not something else.

DrJosh9000 commented on secrets.go:
```diff
@@ -0,0 +1,132 @@
+
+var _ interface {
+	json.Unmarshaler
+	ordered.Unmarshaler
```
Because `Secrets` has a `MarshalYAML` method and we want some assurance that the YAML package will find and use it, we can add `yaml.Marshaler` here:

```suggestion
	ordered.Unmarshaler
	yaml.Marshaler
```

(This _should_ work even though the method takes a `Secrets` as its receiver, and this var is testing interface compatibility with a value of `*Secrets` rather than `Secrets`, because given a non-nil pointer to some data, Go is usually smart enough to dereference it in order to call a method on the non-pointer.)

DrJosh9000 commented on secrets.go:
```diff
@@ -0,0 +1,132 @@
+				secret := Secret{}
+
+				keyVal, _ := ct.Get("key")
+				// Still need two values on the left side to avoid panic
```
This line was a comment for code review explanation purposes more than anything

```suggestion
```

DrJosh9000 reviewed (approved):
This is looking great! Thanks for the effort put into it!

matthewborden committed:
Update secrets.go

Co-authored-by: Josh Deprez <jd@buildkite.com>

matthewborden committed:
Use cmp.Diff more

matthewborden committed:
Update secrets.go

Co-authored-by: Josh Deprez <jd@buildkite.com>

matthewborden committed:
Add support for key/value overloading

Add support for

```
secrets:
  ENV_VAR: SECRET_NAME
```

matthewborden committed:
Add map syntax support and improve marshaling

- Support key-value map syntax: {ENV_VAR: SECRET_KEY}
- Enhance MarshalYAML to use appropriate format (string, map, or object)
- Add comprehensive test coverage for all syntax variants
- Maintain backward compatibility with existing formats
- Follow Google Go style guide

matthewborden committed:
Remove comments

matthewborden committed:
Add test for unmarshaling maps

DrJosh9000 commented on secrets.go:
```diff
@@ -0,0 +1,182 @@
+		// `secrets: null` is invalid - should be omitted entirely or use valid formats
+		return fmt.Errorf("unmarshaling secrets: secrets cannot be null")
+
+	case *ordered.Map[string, interface{}]:
+		// Handle map syntax: {"ENV_VAR": "SECRET_KEY"}
+		return o.Range(func(envVar string, secretKeyVal interface{}) error {
```
The short form of `interface{}` is `any`, which is now pretty standard:

```suggestion
	case *ordered.Map[string, any]:
		// Handle map syntax: {"ENV_VAR": "SECRET_KEY"}
		return o.Range(func(envVar string, secretKeyVal any) error {
```

matthewborden committed:
Replace object format fallback with error

- MarshalYAML now returns error instead of object format
- Prevents invalid configurations from being silently marshaled
- Provides clear error messages for unsupported field combinations
- Add comprehensive test coverage for error conditions
- Update documentation to reflect error behavior

DrJosh9000 commented on secrets.go:
```diff
@@ -0,0 +1,181 @@
+
+func canMarshalAsMap(secrets Secrets) bool {
+	for _, secret := range secrets {
+		if secret.EnvironmentVariable == "" || len(secret.RemainingFields) > 0 {
```
If `secret.EnvironmentVariable == ""`, would it make sense to default it to the value of `secret.Key` for the purposes of marshaling as a map?

matthewborden replied:
Would this effect signing as it's changing the value of the environment variable?

DrJosh9000 replied:
The main thing for signing is to ensure that equivalent objects are marshaled to the same canonical representation. If a `Secrets` can be equivalently represented with a list or with a map, then it would be bad to sign with one and then try to verify with the other. So actually yeah you're probably right.

DrJosh9000 commented on secrets.go:
```diff
@@ -0,0 +1,181 @@
+
+func canMarshalAsSimpleStrings(secrets Secrets) bool {
+	for _, secret := range secrets {
+		if secret.EnvironmentVariable == "" ||
+			secret.Key != secret.EnvironmentVariable ||
+			len(secret.RemainingFields) > 0 {
```
If `secret.Key` cannot be empty, then we don't have to separately check if `secret.EnvironmentVariable` is empty, right?

Either way, we're generally not afraid of longer lines in Go. Some people likely disagree, but I think this isn't too long:


```suggestion
		if secret.EnvironmentVariable == "" || secret.Key != secret.EnvironmentVariable || len(secret.RemainingFields) > 0 {
```

matthewborden committed:
Update secrets.go

Co-authored-by: Josh Deprez <jd@buildkite.com>

matthewborden committed:
Update secrets.go

Co-authored-by: Josh Deprez <jd@buildkite.com>

matthewborden commented on secrets.go:
```diff
@@ -0,0 +1,179 @@
+
+// MarshalYAML returns the most appropriate YAML representation for the secrets.
+// It prioritizes simple string format, then map format, then returns an error for invalid configurations.
+func (s Secrets) MarshalYAML() (interface{}, error) {
```
@moskyb recommended that I use an array of maps for marshalling secrets to avoid locking ourselves out of adding more fields later.

matthewborden committed:
move back to object signitures

matthewborden commented:
Test for existing signed pipelines can be verified with this agent version:

Signed an existing pipeline using buildkite-agent version 3.104.0+10688.2bbc53cb746d46b0e611d7b78df0a5a082a97339

buildkite-agent tool sign unsigned-rspec.yml \
        --jwks-file ~/test/.buildkite/keys/private.json \
        --repo https://github.com/matthewborden/test.git > pipeline.yml

```
steps:
  - label: ":rspec:"
    command: |-
      bundle
      rspec --format progress --format RspecJunitFormatter --out tmp/rspec.xml
    plugins:
      - github.com/buildkite-plugins/docker-buildkite-plugin#v2.0.0:
          image: ruby:2.4
          workdir: /app
    signature:
      algorithm: PS512
      signed_fields:
        - command
        - env
        - matrix
        - plugins
        - repository_url
      value: eyJhbGciOiJQUzUxMiIsImtpZCI6InNlY3JldHMtdGVzdC1rZXkifQ..MyvkRfBNMou4jA2CVyKSqiBzT5-tNKI2E2sqZj9GGBeW6jIzKY4IoLcuZJ_PVwGt2n-QbVgOJEa_ePhNTbESoIz7Gu8amUaqJfi7thgHV9AoeLb0CUPy5viAs-WNmLDv33l92vjoD5DMzC31L8Pblu7-yrarknHZuD__QOTObl3bj21h0FSqn3SFOXbS31WtZeIvc5qF7bsjYLeqBJ9fEXqMGpF0kgcLiLSkAg7K2fOTHxvZDnEEM7XbWUl0derRbOT5hdgRWxYTMcV3-0VShm8reA4OAmlqAGECRKQvV_S0UzQi9Lm_5a7bvv3kKduzEPY_1To8ezfEXb1hjWKUwA
    artifact_paths: tmp/rspec-*.xml
  - continue_on_failure: true
    wait: null
  - label: ":junit:"
    command: .buildkite/junit.sh
    signature:
      algorithm: PS512
      signed_fields:
        - command
        - env
        - matrix
        - plugins
        - repository_url
      value: eyJhbGciOiJQUzUxMiIsImtpZCI6InNlY3JldHMtdGVzdC1rZXkifQ..VaeovEs7iT3xFRliW5FMVR_VXVYlfrqQvdCDHhFuBxPqH5YpDQMuviS5lZ6z9HqiegbvQo8Ayay5VIISzCsE9Shxh3jJyVbyvJgG4B88Rs_zxZ8y5gzDUMAvEOt-04P-NUM8AepTVTALJTCkAxOmcKy6KO7XQiyrmDBB4-08ma83p_UIuhguEeuj53WqauGCkbB9C1Ca1-p3aXtH706OQxxEMfO0njIShE8erCfPjO5975mFOSz1yNp5As2KRwHt0hYpQ-NaXySjMeY0YM6aTKuPRlLHW0uKxvurqlIzxdyggcAWJi_gCAlHvibhQOSV7PajrL9ylR_C5VtsRjI8Sw
  - wait
  - label: ":rocket:"
    command: .buildkite/deploy.sh
    signature:
      algorithm: PS512
      signed_fields:
        - command
        - env
        - matrix
        - plugins
        - repository_url
      value: eyJhbGciOiJQUzUxMiIsImtpZCI6InNlY3JldHMtdGVzdC1rZXkifQ..nobAgqIjdLIspR23PkmIcCvzY8xIUf1Q5vyx5MUYaVFiG844jMkEY3hR6SqypYNacIRZLsXc_DzXvFgSkH6kLWdUvIY0XazCtESbdeSuUvGjwOzwbkimlWG6cbDOCfv-pmiqVOe5N66MGOxtAUImyunVNXkCEtRUP1nV4N0GkybF4yVMxbRJ7byouGulkSUKciFCijJ3zz-WcNKcgiKIawAd-kqKmZY9KNdhDtps07pm0woceQpHZdccekl-t6cDOgdta8FaSFTW4vUianQ1SxuvMQkgi9jubbLxUe0WLIZDx2x_Emr973R1As211kc4mN2yPyg8-ZMsI2TkFbuFKw
```

Verify with buildkite-agent (local) 3.104.0+x.f7ee2b210a01aaf9fb2396b72811151d8086e32c.dirty

```
~/buildkite/agent/buildkite-agent start --debug --tags "queue=default" --token TOKEN --endpoint http://agent.buildkite.localhost/v3 --verification-jwks-file ~/test/.buildkite/keys/public.json --verification-failure-behavior warn --debug-http
```

Successfully verified:
<img width="2043" height="386" alt="Screenshot 2025-09-11 at 11 42 37 am" src="https://github.com/user-attachments/assets/335bf88b-711c-49eb-803a-351e04a89108" />

matthewborden commented:
Testing version signed without secrets with this new agent version:

```
./buildkite-agent tool sign ~/test/.buildkite/unsigned-rspec.yml \
        --jwks-file ~/test/.buildkite/keys/private.json \
        --repo https://github.com/matthewborden/test.git > ~/test/.buildkite/pipeline.yml
```

```
steps:
  - label: ":rspec:"
    command: |-
      bundle
      rspec --format progress --format RspecJunitFormatter --out tmp/rspec.xml
    plugins:
      - github.com/buildkite-plugins/docker-buildkite-plugin#v2.0.0:
          image: ruby:2.4
          workdir: /app
    signature:
      algorithm: PS512
      signed_fields:
        - command
        - env
        - matrix
        - plugins
        - repository_url
      value: eyJhbGciOiJQUzUxMiIsImtpZCI6InNlY3JldHMtdGVzdC1rZXkifQ..Qr7C5IUdMp5QrCbjlrbYDXBonEIiiaDCSqYbbcTyIaz4j1KmwPG36Xpb3hPuMPDMsDy-zKuwfJS7nIwCw2rL-eWcsT9NXTKU-BP1yRgBLqXmrmrZwm1LF7Nm-wHXkmu0KcdqvXUs8SmMHde-oty1XYmPZ90_fw7Nhi9N9yOAcMckEA6s0z0iyENyolph0CwOFC0TKpT7WTQg09VNszuR0GB_YGA9tHgJVERsX9eS8z7MdVddPvR_badLwsWWefrIYQMAu_jt6antRNFkJNaTe95kaXN-A6TtdfemK2-iAiLkS7ApafAoiD0O2OkNE0Hc6zoG5wvqwgasn4H2P6Qg8w
    artifact_paths: tmp/rspec-*.xml
  - continue_on_failure: true
    wait: null
  - label: ":junit:"
    command: .buildkite/junit.sh
    signature:
      algorithm: PS512
      signed_fields:
        - command
        - env
        - matrix
        - plugins
        - repository_url
      value: eyJhbGciOiJQUzUxMiIsImtpZCI6InNlY3JldHMtdGVzdC1rZXkifQ..CBOqnAkdEumAUA87GW723gggrZ-tAi09c6S9mEGKfEPm-5UutG5EP-DOdAQRPe6JKu_eHyo35d2e8UABv4XLc47nLKQz2YyzfLGOOSx4GL4NpbS5dkez4ftOy4MYkswlu3EsF59zVsDl92CA6Od69i2gg6SXGMWvyOP6nnrhS8g-KaovBaGWhM-IlnGUs29x9-7k9tsug7bpARsnIyT8Nr7Q7jJquHSy5wY35VXvOKZbCOJJP50tPXuazzp6vrRitnYIAuahp91bBtFInFLpXgWa3MtGdi19JvFZFJPJ0iqEZeq4f2lF1s4LS6mTfg7t4jigUMlNNCxr6jAGKkZblA
  - wait
  - label: ":rocket:"
    command: .buildkite/deploy.sh
    signature:
      algorithm: PS512
      signed_fields:
        - command
        - env
        - matrix
        - plugins
        - repository_url
      value: eyJhbGciOiJQUzUxMiIsImtpZCI6InNlY3JldHMtdGVzdC1rZXkifQ..LYrToISQaaEcJq70_7thOqb6t9tjBwmtNkktphneUS0R7omiulCQ9osG37ft1pfGOlXQLueRfU7HKDTPjDrT6QbwESNZUopJSrzCeMkdyO1OwZ1qrGXpXj8ovDsw1UK5gWDylSKlguq8HUKBEhgyYSTsLteeMcvD7GByX9o-PJABChZ_RLk0mgRUUBNnA0hjfp3hrN92mr7rReiyzeKome42dez0JZsjTdUlztd5mv1Pc5lDxAdnHC2eO6ZB-g-QbjAZXV3Ks-z1Os3YpfVA43swZaCfxfJOlG_RnvYCJv7JK6CXDmo_cJnlVUBSok6wqGhQ3z7JqOI6yTT7y-YZGQ
```

Running with `buildkite-agent version 3.104.0+x.f7ee2b210a01aaf9fb2396b72811151d8086e32c.dirty`

```
buildkite-agent start --debug --tags "queue=default" --token TOKEN --endpoint http://agent.buildkite.localhost/v3 --verification-jwks-file ~/test/.buildkite/keys/public.json --verification-failure-behavior warn --debug-http
```


<img width="2041" height="381" alt="Screenshot 2025-09-11 at 12 05 04 pm" src="https://github.com/user-attachments/assets/c3e1720a-5460-45b5-910f-370a85f87273" />

The signatures differ between agent versions. @moskyb is that expected now that we've added secrets to the payload? (even though this example doesn't have secrets?).

moskyb commented:
PS512 produces non-deterministic signatures, so signing the same step twice will always produce different signatures. so long as they both verify, it'll probably be fine.

if you wanna double check, `EdDSA` keys produce deterministic signatures.
